### ISSUE TITLE
A length prefix is added to HDSeed before calculating its fingerprint. This is a non-standard ZIP32.

### ISSUE THREAD

**Issue by bigbrain** [zcash/zcash#4641]
Hello.
I don't know if this is a bug or a feature.

The CBLAKE2bWriter class is used to compute the HDSeed fingerprint. But it uses serialization and add length prefix to HDSeed before calculating the hash. This changes the hash and makes it incompatible with ZIP32.

zip32.cpp:
```C++
uint256 HDSeed::Fingerprint() const
{
    CBLAKE2bWriter h(SER_GETHASH, 0, ZCASH_HD_SEED_FP_PERSONAL);
    h << seed;
    return h.GetHash();
}
```

hash.h:
```C++

    // invalidates the object
    uint256 GetHash() {
        uint256 result;
        crypto_generichash_blake2b_final(&state, (unsigned char*)&result, 32);
        return result;
    }

    template<typename T>
    CBLAKE2bWriter& operator<<(const T& obj) {
        // Serialize to this stream
        ::Serialize(*this, obj);
        return (*this);
    }
```

Fox Example:

Now:
**fingerprint = blake2b256(noKey, noSalt, "Zcash_HD_Seed_FP".getBytes(), 0x20|HDSeed)  and then LITTLE-ENDIAN to BIG-ENDIAN**
```
From test wallet dump:
# HDSeed=469d5cd00241515c8ada10e4d22f9afa7206a92f97d51ab1f74f6b2234e01ee4 
fingerprint=468b9c6525b2b52fc12397cc3878d2f6b2e9a07a8eb11bed367ae27a3dd07be5
```
But by ZIP32 should be:
**fingerprint = blake2b256(noKey, noSalt, "Zcash_HD_Seed_FP".getBytes(), HDSeed)**
```
Should be:
HDSeed=469d5cd00241515c8ada10e4d22f9afa7206a92f97d51ab1f74f6b2234e01ee4
fingerprint=195ae4275dd2e46f5f4cba292acc0a3297c919346034fd6c681fe156054ee037
```

**Comment by str4d** [comment]
The length-prefixing was completely unintentional 🤦‍♂️ it was supposed to be handled the same way as deriving the master extended spending key, but because seeds can be any length, they are stored as a `std::vector` in C++. At this point, it's an errata in ZIP 32 that should be fixed there.

Note that the little-endian to big-endian conversion in the hex encoding is not part of the ZIP, but part of `zcashd`'s interface. All 32-byte values are byte-flipped when hex-encoded due to Bitcoin Core doing this for `uint256` values, which we inherited. When decoding the hex 32-byte value, the bytes would be un-flipped, so in binary the outputs would remain the same.

**Comment by daira** [comment]
Fixed, thanks! :-) Please reopen if you think that commit didn't fix it completely (and feel free to continue discussing if needed).

**Comment by str4d** [comment]
@daira the current behaviour in `zcashd` is to prefix with the length of the seed, not a fixed `0x20`. It just currently happens that all `zcashd` seeds are 32 bytes because `HD_WALLET_SEED_LENGTH = 32`.

So, we can either change ZIP 32 to prefix with the seed length, or change it to prefix with a hard-coded `0x20` and then also change `zcashd` to do the same (so we don't have this problem again if `HD_WALLET_SEED_LENGTH` is altered).

**Comment by daira** [comment]
I realised that just after I'd made the other commit :-)

**Comment by bigbrain** [comment]
I left a small comment on the last commit. I think the issue can be closed.

**Comment by daira** [comment]
Off-by-one fixed in 408a0a744c4b19dfbf7ab82f587cf010c1161660
